src/hotspot/share/opto/loopUnswitch.cpp

```diff
@@ -102,114 +99,122 @@ IfNode* PhaseIdealLoop::find_unswitching_candidate(const IdealLoopTree *loop) co
-void PhaseIdealLoop::do_unswitching(IdealLoopTree *loop, Node_List &old_new) {
-  LoopNode* head = loop->_head->as_Loop();
-  if (has_control_dependencies_from_predicates(head)) {
-    return;
+// This class creates an If node (i.e. loop selector) that selects if the true-path- or the false-path-loop should be
+// executed at runtime. This is done by finding an invariant and non-loop-exiting unswitch candidate If node (guaranteed
+// to exist at this point) to perform Loop Unswitching on.
+class UnswitchedLoopSelector : public StackObj {
+  PhaseIdealLoop* const _phase;
+  IdealLoopTree* const _outer_loop;
+  Node* const _original_loop_entry;
+  IfNode* const _unswitch_candidate;
+  IfNode* const _selector;
+  IfTrueNode* const _true_path_loop_proj;
+  IfFalseNode* const _false_path_loop_proj;
+
+  enum PathToLoop { TRUE_PATH, FALSE_PATH };
+
+ public:
+  UnswitchedLoopSelector(IdealLoopTree* loop)
+      : _phase(loop->_phase),
+        _outer_loop(loop->_head->as_Loop()->is_strip_mined() ? loop->_parent->_parent : loop->_parent),
+        _original_loop_entry(loop->_head->as_Loop()->skip_strip_mined()->in(LoopNode::EntryControl)),
+        _unswitch_candidate(find_unswitch_candidate(loop)),
+        _selector(create_selector_if()),
+        _true_path_loop_proj(create_proj_to_loop(TRUE_PATH)->as_IfTrue()),
+        _false_path_loop_proj(create_proj_to_loop(FALSE_PATH)->as_IfFalse()) {
+  }
+  NONCOPYABLE(UnswitchedLoopSelector);
+
+ private:
+  IfNode* find_unswitch_candidate(IdealLoopTree* loop) {
+    IfNode* unswitch_candidate = _phase->find_unswitch_candidate(loop);
+    assert(unswitch_candidate != nullptr, "must exist");
+    assert(_phase->is_member(loop, unswitch_candidate), "must be inside original loop");
+    return unswitch_candidate;
   }
 
-  // Find first invariant test that doesn't exit the loop
-  IfNode* unswitch_iff = find_unswitching_candidate((const IdealLoopTree *)loop);
-  assert(unswitch_iff != nullptr, "should be at least one");
+  IfNode* create_selector_if() const {
+    const uint dom_depth = _phase->dom_depth(_original_loop_entry);
+    _phase->igvn().rehash_node_delayed(_original_loop_entry);
+    BoolNode* unswitch_candidate_bool = _unswitch_candidate->in(1)->as_Bool();
+    IfNode* selector_if = IfNode::make_with_same_profile(_unswitch_candidate, _original_loop_entry,
+                                                         unswitch_candidate_bool);
+    _phase->register_node(selector_if, _outer_loop, _original_loop_entry, dom_depth);
+    return selector_if;
+  }
 
-#ifndef PRODUCT
-  if (TraceLoopOpts) {
-    tty->print("Unswitch   %d ", head->unswitch_count()+1);
-    loop->dump_head();
+  IfProjNode* create_proj_to_loop(const PathToLoop path_to_loop) {
+    const uint dom_depth = _phase->dom_depth(_original_loop_entry);
+    IfProjNode* proj_to_loop;
+    if (path_to_loop == TRUE_PATH) {
+      proj_to_loop = new IfTrueNode(_selector);
+    } else {
+      proj_to_loop = new IfFalseNode(_selector);
+    }
```

Comment:
```suggestion
    IfProjNode* proj_to_loop = (path_to_loop == TRUE_PATH) ? new IfTrueNode(_selector)
                                                             new IfFalseNode(_selector);
```
A question of taste. Just an idea.